encoding/dot/dotparser/internal/astx/astx.go

```diff
@@ -0,0 +1,322 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package astx implements utility functions for generating abstract syntax
+// trees of Graphviz DOT graphs.
+package astx
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gonum/graph/encoding/dot/dotparser/ast"
+	"github.com/gonum/graph/encoding/dot/dotparser/internal/token"
+)
+
+// === [ File ] ================================================================
+
+// NewFile returns a new file based on the given graph.
+func NewFile(graph interface{}) (*ast.File, error) {
```

Comment:
Not at all, the accepted type of `NewFile` will always be `*ast.Graph`. See [dot.bnf](https://raw.githubusercontent.com/mewpull/graph/83ade5cff70af2c164be0fe012f5b47cf65244e1/encoding/dot/dotparser/internal/dot.bnf) for the DOT grammar read by Gocc.

The Gocc parser generator is invoking `astx.NewFile`, when a *reduce* action is triggered for the `File` production rule, which only includes the `Graph` non-terminal; which has the type of the return value from `astx.NewGraph`.

I realize that the `interface{}` arguments makes you feel queasy, I would tend to agree. Unfortunately, this is by design of Gocc, to make it possible for users to write BNF grammars with unknown reduce actions. If the reduce actions were known to Gocc in advance, the reduce function arguments may have more narrow types.

Truely, I don't know if there is a clean way to improve this situation using parser generators in Go. If there is, I would be most interested in discussing how, so we may improve the situation for all Gocc users.

That is the main reason why the Gocc specific and generated packages are put inside an `internal` directory. The API is not intended to be called by anyone, except Gocc.